### PR TITLE
Optimize Spree::PromotionHandler::Cart

### PR DESCRIPTION
**Pull request by mamhoff**
This speeds up the default cart promotion handler by preloading records and taking advantage of already preloaded records where possible. 

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message

### CODE REVIEW

File: core/app/models/spree/promotion_handler/cart.rb [hunk 1]
```diff
@@ -33,7 +33,16 @@ def activate
...
+        promos = connected_order_promotions | sale_promotions
+        preloader = ActiveRecord::Associations::Preloader.new
+        promos.map do |promotion|
+          preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Product" }, :products)
+          preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Store" }, :stores)
+          preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Taxon" }, :taxons)
+          preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::User" }, :users)
+          preloader.preload(promotion.actions.select { |a| a.respond_to?(:calculator) }, :calculator)
+          promotion
+        end
```
Comment (tvdeyen) [inline core/app/models/spree/promotion_handler/cart.rb:45]: TIL :)

**Review by tvdeyen — APPROVED** [review]
Very nice work. This will help a lot of people.

**Review by kennyadsl — COMMENTED** [review]
Nice, left some questions but seems like a good optimization. Do you have a sense of the performance gained out of this?

File: core/app/models/spree/promotion_handler/cart.rb [hunk 2]
```diff
@@ -33,23 +33,37 @@ def activate
...
       end
 
       def connected_order_promotions
-        Spree::Promotion.active.includes(:promotion_rules).
-          joins(:order_promotions).
-          where(spree_orders_promotions: { order_id: order.id }).readonly(false).to_a
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/app/models/spree/promotion_handler/cart.rb:42]: Honestly, I don't know why it's needed. Consider this: 
```
[15] pry(main)> relation = Spree::Promotion.includes(:promotion_rules).
joins(:order_promotions).ions: { order_id: order.id }).readonly(false).to_a.first.readonly?
  Spree::Promotion Load (0.7ms)  SELECT "spree_promotions".* FROM "spree_promotions" INNER JOIN "spree_orders_promotions" ON "spree_orders_promotions"."promotion_id" = "spree_promotions"."id" WHERE "spree_orders_promotions"."order_id" = $1  [["order_id", 5161100]]
  Spree::PromotionRule Load (2.0ms)  SELECT "spree_promotion_rules".* FROM "spree_promotion_rules" WHERE "spree_promotion_rules"."promotion_id" = $1  [["promotion_id", 22]]
=> false
[16] pry(main)> relation = Spree::Promotion.includes(:promotion_rules).
[16] pry(main)> relation = Spree::Promotion.includes(:promotion_rules).
joins(:order_promotions).ions: { order_id: order.id }).readonly(true).to_a.first.readonly?
  Spree::Promotion Load (0.9ms)  SELECT "spree_promotions".* FROM "spree_promotions" INNER JOIN "spree_orders_promotions" ON "spree_orders_promotions"."promotion_id" = "spree_promotions"."id" WHERE "spree_orders_promotions"."order_id" = $1  [["order_id", 5161100]]
  Spree::PromotionRule Load (0.4ms)  SELECT "spree_promotion_rules".* FROM "spree_promotion_rules" WHERE "spree_promotion_rules"."promotion_id" = $1  [["promotion_id", 22]]
=> true
[17] pry(main)> order.promotions.first.readonly?
  Spree::Promotion Load (0.5ms)  SELECT "spree_promotions".* FROM "spree_promotions" INNER JOIN "spree_orders_promotions" ON "spree_promotions"."id" = "spree_orders_promotions"."promotion_id" WHERE "spree_orders_promotions"."order_id" = $1 ORDER BY "spree_promotions"."id" ASC LIMIT $2  [["order_id", 5161100], ["LIMIT", 1]]
=> false
```
`.readonly(false)` makes sure the returned objects are NOT read-only, which `order.promotions` also - although implicitly - does.

File: core/app/models/spree/promotion_handler/cart.rb [hunk 3]
```diff
@@ -33,23 +33,37 @@ def activate
...
+        order.promotions.select(&:active?)
       end
 
       def sale_promotions
-        Spree::Promotion.where(apply_automatically: true).active.includes(:promotion_rules)
+        Spree::Promotion.where(apply_automatically: true).active.includes(promotion_includes)
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/app/models/spree/promotion_handler/cart.rb:53]: Hm, I'm trying to be explicit about pre-loading in this PR. It's not strictly necessary. 
- the `active` scope already preloads promotion actions through a `join`
- `promotion_includes` isn't really used anywhere else. 

However, we're only preloading rules and actions on the activatable promotions; and in our local app I've done the preloading of rules and actions for the order promotions at the controller level. It's arguable though that they're not really needed anywhere but here, and then I should add `includes(promotion_includes)` also to the order promotions...

File: core/app/models/spree/promotion_handler/cart.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kennyadsl) [reply core/app/models/spree/promotion_handler/cart.rb:42]: Reading [here](https://apidock.com/rails/v6.1.3.1/ActiveRecord/QueryMethods/readonly), using `readonly` the relation will be marked as read-only, which means that we can't update the objects in that relation. But, we are setting this to false, as you are pointing out, which is the same as not setting this at all. My first thought is that there is (or most likely was) something that was setting the relation to read-only in some scope and this is undoing that operation. Just for reference, this is where it has been introduced: https://github.com/solidusio/solidus/commit/a5f3796c54a61462e061447e98dd5aea9d7b5382. 

By the way, I'm ok removing this if no spec is failing, thanks!

File: core/app/models/spree/promotion_handler/cart.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (kennyadsl) [reply core/app/models/spree/promotion_handler/cart.rb:53]: If `promotion_actions` are already added by the `active` scope and it's required to keep the method working (eg, if we remove the `active` scope in the future), 👍 being explicit here.

**Comment by mamhoff** [comment]
> Do you have a sense of the performance gained out of this?

I don't really as removing n+1s have peculiar performance characteristics: In isolation, there isn't much in terms of performance improvement you'll see. In our app, this and other n+1 removals (see some of my other PRs) reduced the time for add-to-cart (in which we check stock availability) from up to a minute (!) to consistently under a half a second. However, this was not this PR alone, and dependent on very diligently preloading records in our controllers.

**Review by kennyadsl — APPROVED** [review]

**Review by jarednorman — APPROVED** [review]

**Pull request merged**